sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -1316,6 +1320,13 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
   override def alterPartitions(hive: Hive, tableName: String, newParts: JList[Partition]): Unit = {
     alterPartitionsMethod.invoke(hive, tableName, newParts, environmentContextInAlterTable)
   }
+
+  // HIVE-10319 introduced a new HMS thrift API `get_all_functions` which is used by
+  // `Hive.get` since version 2.1.0, when it loads all Hive permanent functions during
+  // initialization. This breaks compatibility with HMS server of lower versions.
+  // To mitigate here we use `Hive.getWithFastCheck` instead which skips loading the permanent
+  // functions and therefore avoids calling `get_all_functions`.
+  override def getHive(hiveConf: HiveConf): Hive = Hive.getWithFastCheck(hiveConf, false)
```

Comment:
should we make it final? If higher version overrides it accidentally, seems we don't have test to prevent it.

Reply:
Hmm I don't see why ppl would intentionally override this while ignoring the rationale here. On the other hand we might need to override it again in case Hive changes in an incompatible way (say, `Hive.getWithFastCheck` starts to call another newly introduced HMS thrift API).

Reply:
Oh, I thought it may be easily to ignore we have different `getHive` here, when overriding `getHive` on other Shim. If we want to override it again, we at least notice it by compiler error.

Not strong option, anyway. Okay for me as it is.